### DISCUSSION TITLE
Old extension files kept forever

### DISCUSSION THREAD

**Discussion by aptalca** [coder/code-server#3683] [Q&A]
Hi,

I just noticed that the data folder of my code-server instance was getting quite large. Digging in, I realized code-server kept copies of all the old extensions it ever downloaded. There were 2.4GB of python extension files.

Are they not supposed to be pruned when updated? Is this a bug or intended behavior?

If intended, is there a way to prune them without manually going through and deleting the files on disk?

Thanks

**Comment by nooobcoder** [comment]
I guess this is the default behavior of vscode / code @jsjoeio 
I guess code caches old extensions instead of installing again when you want to install a specific version of it. 
As per the discussions there, seems like restarting the code-server container (if docker) or the service shall fix your issue.
https://github.com/microsoft/vscode/issues/106595
In my case, i always faced like the extensions being cached and not cleared at all. 

Let me know if i am wrong in my answer, i am not so old in code-server although, it has been year by now.

**Reply by aptalca** [reply]
Thanks. It looks like that link is about the cached downloads under `CachedExtensionVSIXs` folder. I'm referring to the installed extensions under the `extensions` folder. I had old extension versions going back years even though code server has been restarted plenty of times.

**Reply by jsjoeio** [reply]
I actually noticed this too with old Python extensions hanging around even though I uninstalled them globally. I believe this is intended, but not sure. And I'm not sure if there is any easy way to prune them without doing it manually. Let's see if @code-asher or @oxy know

**Reply by code-asher** [reply]
Yeah it seems that VS Code doesn't actually delete extensions, it just creates a `.obsolete` file with the names of the extensions that shouldn't be loaded.

But I am not sure if they are intended to sit around forever or if there is some cleanup code we are supposed to be running that we aren't (or that we need to implement).
